### PR TITLE
Propose extract_bfs_paths C API

### PR DESCRIPTION
**Pull request by ChuckHastings**
This PR defines the C API for extract_bfs_paths.  Implementation will follow once the API is approved.

Partially addresses #1891

### CODE REVIEW

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by rlratzel** [comment]
rerun tests

Rerunning to get latest updates that should fix build problem.

**Comment by rlratzel** [comment]
I talked with @ChuckHastings offline, and something we discussed is that these new APIs could be used for general path extraction - not just for extracting paths from BFS results - and should be renamed without `_bfs` in the names (and other changes to decouple from BFS, if possible). For example, a general path extraction API could be immediately useful when we add SSSP to the C API, which returns results in the same format.

File: cpp/include/cugraph_c/algorithms.h [hunk 1]
```diff
@@ -194,20 +197,82 @@ void cugraph_bfs_result_free(cugraph_bfs_result_t* result);
...
+  bool_t compute_predecessors,
+  cugraph_bfs_result_t** result,
+  cugraph_error_t** error);
+
+/**
+ * @brief     Opaque extract_bfs_paths result type
+ */
+typedef struct {
+  int align_;
+} cugraph_extract_bfs_paths_result_t;
```
Comment (rlratzel) [inline cpp/include/cugraph_c/algorithms.h:222]: Reminder for either a `FIXME` or an update in this PR to address [this](https://github.com/rapidsai/cugraph/pull/1955#issuecomment-974584672).

**Review by rlratzel — COMMENTED** [review]
Just added a reminder for a `FIXME` we discussed.

Also, this PR could add a unit test too, which calls a stubbed out implementation. I personally would benefit from seeing how the API is meant to be used in the context of an actual BFS call.

**Comment by rlratzel** [comment]
rerun tests

CI failed due to what looks like a problem Jenkins had in accessing github. Rerunning in case this was temporary.

**Review by rlratzel — APPROVED** [review]
Thanks, LGTM. I had some small suggestions which need not hold up my approval.

File: cpp/tests/c_api/extract_paths_test.c [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  cugraph_error_t* ret_error;
+
+  cugraph_resource_handle_t* p_handle                    = NULL;
+  cugraph_graph_t* p_graph                               = NULL;
+  cugraph_paths_result_t* p_paths_result                 = NULL;
+  cugraph_extract_paths_result_t* p_extract_paths_result = NULL;
+  cugraph_type_erased_device_array_t* p_sources          = NULL;
+  cugraph_type_erased_device_array_t* p_destinations     = NULL;
+
+  p_handle = cugraph_create_handle();
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/tests/c_api/extract_paths_test.c:54]: Fixed in latest push

File: cpp/tests/c_api/extract_paths_test.c [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  TEST_ASSERT(test_ret_value, p_handle != NULL, "raft handle creation failed.");
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/tests/c_api/extract_paths_test.c:55]: Fixed in latest push

**Comment by BradReesWork** [comment]
rerun tests

File: cpp/include/cugraph_c/algorithms.h [hunk 4]
```diff
@@ -136,45 +138,50 @@ cugraph_error_code_t cugraph_personalized_pagerank(
...
 
 /**
- * @brief     Opaque bfs result type
+ * @brief     Opaque paths result type
+ *
+ * Store the output of BFS or SSSP, computing predecessors and distances
+ * from a seed.
  */
 typedef struct {
   int align_;
```
Comment (seunghwak) [inline cpp/include/cugraph_c/algorithms.h:147]: Should we better use int32_t instead of int if the intention here is to really enforce 4 byte alignment?

**Review by seunghwak — APPROVED** [review]
Looks good to me, I have just a minor comment to think about.

File: cpp/include/cugraph_c/algorithms.h (hunk 4 above)
Comment (ChuckHastings) [reply cpp/include/cugraph_c/algorithms.h:147]: Good point.  Will fix in the implementation PR.

**Comment by ChuckHastings** [comment]
@gpucibot merge

**Review by Ethyling — APPROVED** [review]

**Pull request merged**